### PR TITLE
 Flush cache after the current transaction is committed (#296)

### PR DESCRIPTION
**Pull request by dtao**
This builds off the work in pull request #299 and fixes the issue raised in #296.

Note: review #299 before reviewing this.

### CODE REVIEW

File: waffle/tests/test_testutils.py
Comment (dtao) [inline waffle/tests/test_testutils.py:6]: Django's `TestCase` class automatically wraps each test method in a transaction and rolls back the transaction after the method returns. Since updating models now doesn't flush the cache until after the transaction is committed, this breaks the logic in most of these tests. Inheriting from `TransactionTestCase` addresses the problem by removing Django's magical auto-rollback behavior, the trade-off being that the tests are a tad slower.

**Comment by dtao** [comment]
I suppose you could technically merge this before merging #299; there would just effectively be no test coverage since the test added in this PR is skipped when running tests with sqlite.

File: waffle/tests/test_waffle.py
Comment (jsocol) [inline waffle/tests/test_waffle.py]: I appreciate all the comments here, this is definitely not an obvious issue.

Would you mind adding a slightly longer description to the doc block (as a second paragraph) and include a link to #296?

Should we have similar tests for `Flag` and `Sample`? I know the behavior exists in `BaseModel` but that's an implementation detail, not a public API. If we needed to refactor it, especially in support of extensible models, it could change or break.

**Review by jsocol — APPROVED** [review]
Again, thanks @dtao! I think we should wait and merge #299 first then get this one in—I'd rather watch the new CI build on master succeed without this patch first.

File: waffle/tests/test_waffle.py
Comment (dtao) [reply waffle/tests/test_waffle.py]: For sure, I can add another comment explicitly referencing #296.

I'm also happy to add equivalent tests for `Flag` and `Sample`.

**Comment by dtao** [comment]
@jsocol Per your suggestions I've added an explanatory comment to the somewhat complex test method to cover this case and also updated the test code to cover flags, switches, and samples.

I squashed this work into one commit because I _think_ that's what you prefer, though I acknowledge that can make it hard to tell what changed since you last reviewed the code. If I've misunderstood, and for future reference I should not squash commits in pull requests that are under review, let me know.

**Comment by dtao** [comment]
Hey @jsocol @clintonb just making sure: with #299 merged, is this PR waiting on me for anything right now? I'm happy to wait patiently; I just don't want to be neglectful if there's something I should be doing to get this PR merged that I've overlooked.

**Review by clintonb — APPROVED** [review]

**Comment by dtao** [comment]
Is there any interest in releasing a new version with this fix?

We are just depending on a fork at the moment. It's always nice to return to the fold when we can :)

**Comment by clintonb** [comment]
I can take the steps to start a release, but I can't actually merge the requisite changes without @jsocol. I've been trying to get merge rights, but I haven't received a response on #278.

I'd recommend either sticking with your fork, or using the latest commit on master for the time being, @dtao.

**Pull request merged**
